Update: Developing and debugging services locally (#30472)
kubernetes/website: content/en/docs/tasks/debug-application-cluster/local-debugging.md

title: Developing and debugging services locally
content_type: task
---
 
<!-- overview -->
 

> Kubernetes applications usually consist of multiple, separate services, each running in its own container. Developing and debugging these services on a remote Kubernetes cluster can be cumbersome, requiring you to [get a shell on a running container](https://kubernetes.io/docs/tasks/debug-application-cluster/get-shell-running-container/) and run your tools inside that remote shell.

Reviewer: How do you feel about:
```suggestion
Kubernetes applications usually consist of multiple, separate services, each running in its own container. Developing and debugging these services on a remote Kubernetes cluster can be cumbersome, requiring you to [get a shell on a running container](https://kubernetes.io/docs/tasks/debug-application-cluster/get-shell-running-container/) in order to run your tools inside that remote shell.
```
Just a nit

Author: Yeah, this looks great to me. I will push a commit for it.

> Kubernetes applications usually consist of multiple, separate services, each running in its own container. Developing and debugging these services on a remote Kubernetes cluster can be cumbersome, requiring you to [get a shell on a running container](https://kubernetes.io/docs/tasks/debug-application-cluster/get-shell-running-container/) in order to run your tools inside that remote shell.

Reviewer 2: ```suggestion
Kubernetes applications usually consist of multiple, separate services, each running in its own container. Developing and debugging these services on a remote Kubernetes cluster can be cumbersome, requiring you to [get a shell on a running container](/docs/tasks/debug-application-cluster/get-shell-running-container/) in order to run your tools inside that remote shell.
```

We use reference style links with Hugo instead of direct links everywhere, could you please revert this? 🙂

Reviewer 3: +1 to this

Author: Thanks for the feedback, @RinkiyaKeDad & @divya-mohan0209. I just pushed a fix for this.

 
## Developing or debugging an existing service
 
When developing an application on Kubernetes, you typically program or debug a single service. The service might require access to other services for testing and debugging. One option is to use the continuous deployment pipeline, but even the fastest deployment pipeline introduces a delay in the program or debug cycle.
 
Use the `telepresence intercept $SERVICE_NAME --port $LOCAL_PORT:REMOTE_PORT` command to create an "intercept" for rerouting remote service traffic.
 
Where:
 
-   `$SERVICE_NAME`  is the name of your local service
-   `$LOCAL_PORT` is the port that your service is running on your local workstation
-   And `$REMOTE_PORT` is the port your service listens to in the cluster
 
Running this command will tell Telepresence to send remote traffic to your local service instead of the service in the remote Kubernetes cluster.  With this, you can then make edits to your service source code locally, save, and see the corresponding changes when accessing your remote application take effect immediately. You can also run your local service using a debugger or any other local development tool.
 

> <!-- how telepresence works -->

Reviewer 2: I believe we can remove this since the heading says the same :)

Author: Are you suggesting I remove the entire section or just <!-- how telepresence works --> ? @RinkiyaKeDad

Reviewer 2: I mean just the comment `<!-- how telepresence works -->` 🙂

Author: Fixed, @RinkiyaKeDad

 
## How does Telepresence work?
 
Telepresence installs a traffic-agent sidecar next to your existing application's container running in the remote cluster. It then captures all traffic requests going into the Pod, and instead of forwarding this to the application in the remote cluster, it routes all traffic (when you create a [global intercept](https://www.getambassador.io/docs/telepresence/latest/concepts/intercepts/#global-intercept)) or a subset of the traffic (when you create a [personal intercept](https://www.getambassador.io/docs/telepresence/latest/concepts/intercepts/#personal-intercept)) to your local development environment.
 
## {{% heading "whatsnext" %}}
 
If you're interested in a hands-on tutorial, check out [this tutorial](https://blog.jetbrains.com/idea/2021/05/easily-debug-java-microservices-running-on-kubernetes-with-intellij-idea/) that walks through locally debugging Java microservices with IntelliJ IDEA and Telepresence connected to a remote Kubernetes cluster.

>  

Reviewer 2: I'm not sure if we need this in the Kubernetes documentation since we're already linking to the Telepresence website if folks want to learn more. @sftim @reylejano WDYT?

Author: Thanks for the feedback @RinkiyaKeDad. If you check the original [document](https://kubernetes.io/docs/tasks/debug-application-cluster/local-debugging/#what-s-next), you will see it actually includes a tutorial about using Telepresence with Google Kubernetes Engine's Guestbook applications. I had initially replaced it with another article because it was outdated. 

Now, that I have sent a PR to the Google Cloud team, I have reverted this document to the original article since it will be up to date when my PR is merged by the Google Cloud team. Let me know what you think.

Author: Hello @sftim and @reylejano. Can you please share your thoughts on this? So, I can make the necessary changes (if any) and get the pull request merged. Thank you. 🙏

---

<!-- overview -->
 

> Kubernetes applications usually consist of multiple, separate services, each running in its own container. Developing and debugging these services on a remote Kubernetes cluster can be cumbersome, requiring you to [get a shell on a running container](/docs/tasks/debug-application-cluster/get-shell-running-container/) in order to run your tools inside that remote shell.

Reviewer 4: ```suggestion
Kubernetes applications usually consist of multiple, separate services, each running in its own container. Developing and debugging these services on a remote Kubernetes cluster can be cumbersome, requiring you to [get a shell on a running container](/docs/tasks/debug-application-cluster/get-shell-running-container/) in order to run debugging tools.
```

Reviewer 4: I think there still might be a better way to rephrase this, open to alternatives. In the current state, it doesn't read clearly (to me).

> `telepresence` is a tool used to ease the process of developing and debugging services locally while proxying the service to a remote Kubernetes cluster. Using `telepresence` allows you to use custom tools, such as a debugger and IDE, for a local service and provides the service full access to ConfigMap, secrets, and the services running on the remote cluster.

Reviewer 4: ```suggestion
`telepresence` is a tool to ease the process of developing and debugging services locally while proxying the service to a remote Kubernetes cluster. Using `telepresence` allows you to use custom tools, such as a debugger and IDE, for a local service and provides the service full access to ConfigMap, secrets, and the services running on the remote cluster.
```

https://kubernetes.io/docs/contribute/style/style-guide/#use-present-tense

 
This document describes using `telepresence` to develop and debug services running on a remote cluster locally.

## {{% heading "prerequisites" %}}

* Kubernetes cluster is installed
* `kubectl` is configured to communicate with the cluster
* [Telepresence](https://www.telepresence.io/docs/latest/install/) is installed

<!-- steps -->

## Connecting your local machine to a remote Kubernetes cluster
 
After installing `telepresence`, run `telepresence connect` to launch it's Daemon and connect your local workstation to the cluster.

```
$ telepresence connect
 
Launching Telepresence Daemon
...
Connected to context default (https://<cluster public IP>)
```

> You are now connected to the network of your remote cluster and so you can curl services using the Kubernetes syntax e.g. `curl -ik https://kubernetes.default`

Reviewer 4: ```suggestion
You can curl services using the Kubernetes syntax e.g. `curl -ik https://kubernetes.default`
```

 
## Developing or debugging an existing service

When developing an application on Kubernetes, you typically program or debug a single service. The service might require access to other services for testing and debugging. One option is to use the continuous deployment pipeline, but even the fastest deployment pipeline introduces a delay in the program or debug cycle.
 
Use the `telepresence intercept $SERVICE_NAME --port $LOCAL_PORT:REMOTE_PORT` command to create an "intercept" for rerouting remote service traffic.
 
Where:

-   `$SERVICE_NAME`  is the name of your local service
-   `$LOCAL_PORT` is the port that your service is running on your local workstation
-   And `$REMOTE_PORT` is the port your service listens to in the cluster

> Running this command will tell Telepresence to send remote traffic to your local service instead of the service in the remote Kubernetes cluster.  With this, you can then make edits to your service source code locally, save, and see the corresponding changes when accessing your remote application take effect immediately. You can also run your local service using a debugger or any other local development tool.

Reviewer 4: ```suggestion
Running this command tells Telepresence to send remote traffic to your local service instead of the service in the remote Kubernetes cluster. Make edits to your service source code locally, save, and see the corresponding changes when accessing your remote application take effect immediately. You can also run your local service using a debugger or any other local development tool.
```
